dao/mongodb.rya/src/main/java/org/apache/rya/mongodb/aggregation/PipelineResultIteration.java

```
@@ -0,0 +1,132 @@
+ * {@link AggregationPipelineQueryNode} into {@link BindingSet}s.
+ */
+public class PipelineResultIteration implements CloseableIteration<BindingSet, QueryEvaluationException> {
+    private static final int BATCH_SIZE = 1000;
+    private static final ValueFactory VF = ValueFactoryImpl.getInstance();
+
+    private final MongoCursor<Document> cursor;
+    private final Map<String, String> varToOriginalName;
+    private final BindingSet bindings;
+    private BindingSet nextSolution = null;
+
+    /**
+     * Constructor.
+     * @param aggIter Iterator of documents in AggregationPipelineQueryNode's
+     *  intermediate solution representation.
+     * @param varToOriginalName A mapping from field names in the pipeline
+     *  result documents to equivalent variable names in the original query.
+     *  Where an entry does not exist for a field, the field name and variable
+     *  name are assumed to be the same.
+     * @param bindings A partial solution. May be empty.
+     */
+    public PipelineResultIteration(AggregateIterable<Document> aggIter,
+            Map<String, String> varToOriginalName,
+            BindingSet bindings) {
+        aggIter.batchSize(BATCH_SIZE);
+        this.cursor = aggIter.iterator();
+        this.varToOriginalName = varToOriginalName;
+        this.bindings = bindings;
+        lookahead();
+    }
+
+    private void lookahead() {
+        while (nextSolution == null && cursor.hasNext()) {
+            nextSolution = docToBindingSet(cursor.next());
+        }
+    }
+
+    @Override
+    public boolean hasNext() throws QueryEvaluationException {
+        lookahead();
+        return nextSolution != null;
+    }
+
+    @Override
+    public BindingSet next() throws QueryEvaluationException {
+        lookahead();
+        BindingSet solution = nextSolution;
+        nextSolution = null;
+        return solution;
+    }
+
+    @Override
+    public void remove() throws QueryEvaluationException {
```

This should throw an unsupported exception. Remove means you're removing the previously returned object from the underlying collection. This doesn't do that.

---

Agreed, done.